File: linked-list/MyLinkedList.dart
// dart MyLinkedList.dart

class Node<U> {
    Node? next;
    U? value;

    // constructor
    Node(U value) {
        this.value = value;
    }
}

class LinkedList<T> {
    Node? head;

    void addLast(T value) {
        var newNode = Node(value);

        if (this.head == null) {
            this.head = newNode;
        } else {
            var current = this.head;

            while (current?.next != null) {
                current = current?.next;
            }

            current?.next = newNode;
        }
    }

    void addFirst(T value) {
        var newNode = Node(value);

        if (this.head == null) {
            this.head = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
    }

    void removeFirst() {
        if (this.head != null) {
            if (this.head?.next == null) {
                this.head = null;
            } else {
                this.head = this.head?.next;
            }
        }
    }

    void removeLast() {
        if (this.head != null) {
            if (this.head?.next == null) {
                this.head = null;
            } else {
                var current = this.head;

                while (current?.next?.next != null) {
                    current = current?.next;
                }

                current?.next = null;
            }
        }
    }

    T? getFirst() {
        if (this.head != null) {
            return this.head?.value;
        }

        return null;
    }

    T? getLast() {
        if (this.head != null) {
            var current = this.head;

            while (current?.next != null) {
                current = current?.next;
            }

            return current?.value;
        }

        return null;
    }

    void display() {
        if (this.head != null) {
            var current = this.head;

            while (current != null) {
                print("[${current.value}]->");

                current = current.next;
            }
        }
    }
}

void main() {
    // int linked list
    var ll = LinkedList<int>();

    ll.addLast(4);
    ll.addLast(3);
    ll.addLast(2);
    ll.addLast(1);
    ll.addFirst(5);
    ll.addFirst(6);
    ll.addFirst(7);

    ll.removeFirst();
    ll.removeLast();

    ll.display();

    print("First Item: ${ll.getFirst()}");
    print("Last Item: ${ll.getLast()}");

    // string linked list
    var ll2 = LinkedList<String>();

    ll2.addLast("cat");
    ll2.addLast("dog");
    ll2.addFirst("alligator");

    ll2.display();

    print("First Item: ${ll2.getFirst()}");
    print("Last Item: ${ll2.getLast()}");
}

File: linked-list/MyLinkedList.kt
// kotlinc MyLinkedList.kt -include-runtime -d MyLinkedList.jar && java -jar MyLinkedList.jar

// declare variables with var for mutable, or val for non-mutable

// Can have a primary constructor like this, but also many secondary constructors
class Node<V> constructor(var value: V) {
    // instance variables must be initialized, can be done via constructors too
    var next: Node<V>? = null
}

// Generics are supported
class LinkedList<T> {
    private var head: Node<T>? = null // supports nullable types (optionals)

    // when functions don't return anything we can use Unit (like void)
    // though setting a return type is optional if nothing is returned
    fun addLast(value: T): Unit {
        val newNode = Node<T>(value)

        // for instance variables can also be refer with this. i.e this.head
        if (head == null) {
            head = newNode
        } else {
            var current = head

            while (current?.next != null) {
                current = current.next
            }

            current?.next = newNode
        }
    }

    fun addFirst(value: T) {
        val newNode = Node<T>(value)

        if (head == null) {
            head = newNode
        } else {
            newNode.next = head
            head = newNode
        }
    }

    fun removeFirst() {
        if (head != null) {
            if (head?.next != null) {
                head = head?.next
            } else {
                head = null;
            }
        }
    }

    fun removeLast() {
        if (head != null) {
            if (head?.next != null) {
                var current = head

                while (current?.next?.next != null) {
                    current = current.next
                }

                current?.next = null
            } else {
                head = null
            }
        }
    }

    fun getFirst(): T? {
        if (head != null) {
            return head?.value;
        }

        return null
    }

    fun getLast(): T? {
        if (head != null) {
            var current = head

            while (current?.next != null) {
                current = current.next
            }

            return current?.value
        }

        return null;
    }

    fun display() {
        if (head != null) {
            var current = head

            while (current != null) {
                print("[${current.value}]->") // without no new line

                current = current.next
            }

            println() // new line
        }
    }
}

fun main() {
    // string linkedlist
    val ll = LinkedList<String>();

    ll.addLast("cat");
    ll.addLast("dog");
    ll.addFirst("alligator");

    ll.display();

    println("First Item: ${ll.getFirst()}");
    println("Last Item: ${ll.getLast()}");

    // int linkedlist
    val ll2 = LinkedList<Int>();

    ll2.addLast(4);
    ll2.addLast(3);
    ll2.addLast(2);
    ll2.addLast(1);
    ll2.addFirst(5);
    ll2.addFirst(6);
    ll2.addFirst(7);

    ll2.removeFirst();
    ll2.removeLast();

    ll2.display();

    println("First Item: ${ll2.getFirst()}");
    println("Last Item: ${ll2.getLast()}");
}
